File: sdks/flutter/lib/src/resume/cache_resume.dart
import 'dart:convert';

import '../key/cache_key.dart';
import '../key/canonical_json.dart';

class ResumeState {
  ResumeState({
    required this.widgetId,
    required this.stateVersion,
  }) {
    if (widgetId.trim().isEmpty) {
      throw ArgumentError.value(widgetId, 'widgetId', 'must be a non-empty string');
    }
    if (stateVersion < 0) {
      throw ArgumentError.value(stateVersion, 'stateVersion', 'must be a non-negative integer');
    }
  }

  final String widgetId;
  final int stateVersion;
}

class InMemoryResumeStateStore {
  final Map<String, int> _stateByWidget = <String, int>{};

  void update(ResumeState state) {
    final current = _stateByWidget[state.widgetId];
    if (current == null || state.stateVersion > current) {
      _stateByWidget[state.widgetId] = state.stateVersion;
    }
  }

  Map<String, int> snapshot() {
    final keys = _stateByWidget.keys.toList()..sort();
    return {for (final k in keys) k: _stateByWidget[k]!};
  }

  void clear() => _stateByWidget.clear();

  int size() => _stateByWidget.length;
}

Map<String, int> _canonicalizeWidgetStateMap(Map<String, int> widgetStateMap) {
  final keys = widgetStateMap.keys.toList()..sort();
  final canonical = <String, int>{};
  for (final key in keys) {
    if (key.trim().isEmpty) {
      throw ArgumentError.value(key, 'widgetStateMap key', 'must be a non-empty string');
    }
    final value = widgetStateMap[key]!;
    if (value < 0) {
      throw ArgumentError.value(value, 'widgetStateMap[$key]', 'must be a non-negative integer');
    }
    canonical[key] = value;
  }
  return canonical;
}

String buildResumeTokenMaterial({
  required String traceId,
  required Map<String, int> widgetStateMap,
}) {
  if (traceId.trim().isEmpty) {
    throw ArgumentError.value(traceId, 'traceId', 'must be a non-empty string');
  }
  final canonicalMap = _canonicalizeWidgetStateMap(widgetStateMap);
  return '$traceId|${canonicalJsonStringify(canonicalMap)}';
}

Map<String, Object> buildResumeHint({
  required String traceId,
  required Map<String, int> widgetStateMap,
  HashFn? hashFn,
}) {
  if (traceId.trim().isEmpty) {
    throw ArgumentError.value(traceId, 'traceId', 'must be a non-empty string');
  }
  final canonicalMap = _canonicalizeWidgetStateMap(widgetStateMap);
  final result = <String, Object>{
    'widget_state_map': canonicalMap,
  };
  if (hashFn != null) {
    final tokenMaterial = buildResumeTokenMaterial(
      traceId: traceId,
      widgetStateMap: canonicalMap,
    );
    result['resume_token'] = hashFn(utf8.encode(tokenMaterial));
  }
  return result;
}

File: sdks/javascript/resume/cache.resume.ts
import { deterministicSerialize } from "../consistency";
import type { H57HashFunction } from "../key";

const textEncoder = new TextEncoder();

export interface ResumeStateRecord {
  widget_id: string;
  state_version: number;
}

export type WidgetStateMap = Record<string, number>;

export interface ResumeStateStore {
  update(record: ResumeStateRecord): void;
  snapshot(): WidgetStateMap;
  clear(): void;
  size(): number;
}

export interface ResumeHint {
  widget_state_map: WidgetStateMap;
  resume_token?: string;
}

function assertNonEmpty(value: string, field: string): void {
  if (!value?.trim()) {
    throw new TypeError(`${field} must be a non-empty string`);
  }
}

function normalizeStateVersion(value: number): number {
  if (!Number.isFinite(value) || value < 0 || !Number.isInteger(value)) {
    throw new TypeError("state_version must be a non-negative integer");
  }
  return value;
}

function canonicalizeWidgetStateMap(map: WidgetStateMap): WidgetStateMap {
  const out: WidgetStateMap = {};
  const keys = Object.keys(map).sort((a, b) => a.localeCompare(b));
  for (const key of keys) {
    assertNonEmpty(key, "widget_id");
    const value = map[key];
    out[key] = normalizeStateVersion(value);
  }
  return out;
}

export class InMemoryResumeStateStore implements ResumeStateStore {
  private readonly map = new Map<string, number>();

  update(record: ResumeStateRecord): void {
    assertNonEmpty(record.widget_id, "widget_id");
    const next = normalizeStateVersion(record.state_version);
    const current = this.map.get(record.widget_id);
    if (current === undefined || next > current) {
      this.map.set(record.widget_id, next);
    }
  }

  snapshot(): WidgetStateMap {
    const out: WidgetStateMap = {};
    const keys = Array.from(this.map.keys()).sort((a, b) => a.localeCompare(b));
    for (const key of keys) {
      const value = this.map.get(key);
      if (value !== undefined) {
        out[key] = value;
      }
    }
    return out;
  }

  clear(): void {
    this.map.clear();
  }

  size(): number {
    return this.map.size;
  }
}

/**
 * Deterministic token material aligned with H57(trace_id + widget_state_map).
 */
export function buildResumeTokenMaterial(traceId: string, widgetStateMap: WidgetStateMap): string {
  assertNonEmpty(traceId, "trace_id");
  const canonicalMap = canonicalizeWidgetStateMap(widgetStateMap);
  return `${traceId}|${deterministicSerialize(canonicalMap)}`;
}

export function buildResumeHint(input: {
  traceId: string;
  widgetStateMap: WidgetStateMap;
  h57Hash?: H57HashFunction;
}): ResumeHint {
  assertNonEmpty(input.traceId, "trace_id");
  const widgetStateMap = canonicalizeWidgetStateMap(input.widgetStateMap);
  if (!input.h57Hash) {
    return {
      widget_state_map: widgetStateMap
    };
  }

  const material = buildResumeTokenMaterial(input.traceId, widgetStateMap);
  const resumeToken = input.h57Hash(textEncoder.encode(material));

  return {
    widget_state_map: widgetStateMap,
    resume_token: resumeToken
  };
}
